src/java.net.http/share/classes/jdk/internal/net/http/Exchange.java

```diff
@@ -156,20 +157,26 @@ void connection(HttpConnection connection) {
                     this.closeRequested = false;
                 }
             }
-            if (closeRequested) closeConnection(connection);
+            if (closeRequested) closeConnection(connection, cause);
         }
 
-        void closeConnection() {
+        void closeConnection(Throwable error) {
             HttpConnection connection;
+            Throwable cause;
             synchronized (this) {
+                cause = this.cause;
+                if (cause == null) {
+                    cause = this.cause = error;
+                }
                 connection = this.connection;
                 if (connection == null) {
                     closeRequested = true;
                 } else {
                     this.connection = null;
+                    this.cause = null;
```

Comment:
Hello Daniel, I had to read this a few times to understand why we are resetting `this.cause` which we potentially set a few lines above to `null`. From what I understand, we only want to set `this.cause` when the `connection` is not `null`. Do you think this code could instead of written as:

```java
Throwable cause;
synchronized (this) {
 cause = this.cause;
 if (cause == null) {
     cause = error;
 }
 connection = this.connection;
 if (connection == null) {
     closeRequested = true;
     this.cause = cause;
 } else {
     this.connection = null;
     this.cause = null;
 }
}
```

Reply:
Much better yes. I'll steal it!